Add a tripartite connection rule for neuron-astrocyte connectivity

HanjiaJiang opened the pull request:
We implemented a new connection rule, currently named "pairwise_bernoulli_astro", for the creation of tripartite connectivity in neuron-astrocyte networks. This rule is adapted from the "symmetric_pairwise_bernoulli" rule.

The conn_spec parameter "p" determines connection probability between neurons as in the  "symmetric_pairwise_bernoulli" rule. Three new conn_spec parameters are added: "p_syn_astro" determines the probability of each created neuron-neuron connection to be paired with an astrocyte, "max_astro_per_target" determines the maximal number of astrocytes each target neuron can be connected with, and "astro_pool_by_index" determines whether the astrocytes are paired with neurons randomly or by index.

The new syn_spec parameters "weight_pre2post", "weight_pre2astro", "weight_astro2post", "delay_pre2post", "delay_pre2astro", and "delay_astro2post" determine the weights and delays in the tripartite connectivity.

The naming and design in this new rule are still open to discussion and changes. We will continue to improve the code during the review process.

heplesser commented:
This PR introduces a new concept, tripartite connectivity between neurons and astrocytes. The user interface included in the PR is specific to astrocytes. I propose a more generic API for tripartite connections and would like to discuss details here.

## The connection rule

In tripartite connections, we have
- source population
- target population
- third-party population

Exactly how these are connected is up to the specific connection rule. The first connection rule propose here, works as follows:
1. Perform a Bernoulli experiment to test whether to connect a source-target pair. Connect that pair.
2. If a source-target pair is to be connected, perform a second Bernoulli experiment to determine if the third party shall be connected.
3. If the third party is to be connected
    1. Build a pool of third-party nodes specific to the target node chosen in step 1 (see below for details).
    2. Randomly select a single third-party node from the pool.
    3. Connect the source node from step 1 to the chosen third-party node and the third-party node to the target node.

The pool from 3.i has the following properties:
1. For a given target, the same pool is used for all sources connected to this target.
2. The user can specify a pool size, between one and the size of the third-party population.
3. If no pool size is given, the entire third-party population is the pool.
4. Nodes of the third-party population can either be chosen at random (selection with uniform probability without replacement) or based on a regular pattern (for pool size n, assign the first n astrocytes as pool to target 1, the next n to target 2, etc).

I suggest to call this rule `tripartite_bernoulli_with_pool`.  

## API proposal

I suggest here two variants. One maintains the single `Connect()` function, the other introduces a new function.
Before discussing the `Connect()` function, let's look at synapse specifications first.

### Synapse specifications

For tripartite connections, we need synapse specifications for
1. the source -> target connection (the primary connection)
2. the source -> third-party connection (third-party input connection)
3. the third-party -> target connection (third-party output connection)

This is easily achieved by allowing a dictionary of connection specifications, which in turn can contain anything that is legal as a `syn_spec` argument otherwise and defaults to static synapse with weight and delay 1 if missing:
```
    syn_spec={
        "primary": {"model": "tsodyks_synapse", "weight": 1.0, "delay": 1.0},
        "third_in": {"model": "tsodyks_synapse", "weight": 1.0, "delay": 1.0},
        "third_out": {"model": "sic_connection", "weight": 1.0, "delay": 1.0},
    }
```

### Extending `Connect()` semantics

Change from
```python
nest.Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=False)
```
to
```python
nest.Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=False, third=None)
```
allowing calls of the form
```python
nest.Connect(pre_neurons, post_neurons, <conn spec>, <syn spec>, third=astrocytes)
nest.Connect(pre_neurons, post_neurons, third=astrocytes, conn_spec={...}, syn_spec_={...})
```
The downside of this approach is that one either needs to place the third-party argument last or use argument names for `conn_spec` and `syn_spec`.

### Introducing a new connection function


```python
nest.TripartiteConnect(pre, post, third, conn_spec=None, syn_spec=None, return_synapsecollection=False)
```
called as
```python
nest.TripartiteConnect(pre_neurons, post_neurons, astrocytes, <conn spec>, <syn spec>)
```
This is slightly more expressive and would use the same `syn_spec` semantics as the extended `Connect()` alternative, see below. Alternatively, one could have all syn specs explicit:

```python
nest.TripartiteConnect(pre, post, third, conn_spec=None, *, 
  syn_spec_primary=None, syn_spec_third_in=None, syn_spec_third_out=None, return_synapsecollection=False)
```
called as
```python
nest.TripartiteConnect(pre_neurons, post_neurons, astrocytes, <conn spec>, 
   syn_spec_primary={...}, syn_spec_third_in={...}, syn_spec_third_out={...})
```
In this case, I believe one should force named keyword args for the syn specs to avoid confusion.

### Connection specifications

These will depend on the connection rule for details. For the `tripartite_bernoulli_with_pool`, I suggest

```python
    conn_spec={
        "rule": "tripartite_bernoulli_with_pool",
        "p_primary": 0.2,
        "p_cond_third": 0.5,
        "pool_size": 3,
        "random_pool": False,
    }
```
where `p_primary` is the probability to create a source-target connection and `p_cond_third` the probability to create a third-party connection if a primary connection is created.


## Open questions
1. If we all `Connect()` twice on the same target and third party populations, do we required the same random pools or may the pools be different from call to call?
5. Shall we restrict index-based pooling to cases where third-party, target and pool size are commensurable? If we do not restrict like this, how do we handle nodes that do not fit?

HanjiaJiang commented:
Thank you for the comments!

For question 1, so far I cannot think of a reason to restrict each neuron to the same astrocyte pool for a second or even more call of Connect().

For question 2, to me it seems like a reasonable solution, to allow index-based pooling only if third-party, target and pool size are commensurable. This way we can minimize the effort on solving the asymmetry in the pooling. To deal with the borders (begin and end of population), we probably still need the current approach or a wrap-around approach, but the asymmetry can be mostly (if not all) avoided either way, if the proposed restriction is applied.

HanjiaJiang commented:
For the proposal in general, in our last NEST meeting (Sep. 25) it seems
```python
nest.Connect(pre_neurons, post_neurons, <conn spec>, <syn spec>, third=astrocytes)
```
is preferred, which I also agree. Please remind me if I am wrong, thanks.
I am only wondering if "p_cond_third" might be a little confusing because we use "cond" for conductance in some other places in NEST.
Thanks again for the proposal!

heplesser commented:
@HanjiaJiang I just sent you a PR with updates for the remaining issues. Once that is merged and appears here, could you click the "Ready for review" button above?

@jessica-mitchell Could you review the documentation and examples? One point I am unsure about is that the [astrocyte_small_network.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_small_network.py) and the [astrocyte_brunel.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_brunel.py) write figures to files in subdirectories they create. This does not seem to be in line with normal practice for examples. Should they display figures with `plt.show()`?  And will all work if a script produces more than one figure?

@HanjiaJiang and @jessica-mitchell I added an explanation of the tripartite rule to `connection_management.rst`, but I think the discussion of the pools would be much clearer with a figure. Could you to collaborate on getting an illustrative figure in?

HanjiaJiang commented:
> @HanjiaJiang I just sent you a PR with updates for the remaining issues. Once that is merged and appears here, could you click the "Ready for review" button above?
> 
> @jessica-mitchell Could you review the documentation and examples? One point I am unsure about is that the [astrocyte_small_network.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_small_network.py) and the [astrocyte_brunel.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_brunel.py) write figures to files in subdirectories they create. This does not seem to be in line with normal practice for examples. Should they display figures with `plt.show()`? And will all work if a script produces more than one figure?
> 
> @HanjiaJiang and @jessica-mitchell I added an explanation of the tripartite rule to `connection_management.rst`, but I think the discussion of the pools would be much clearer with a figure. Could you to collaborate on getting an illustrative figure in?

Thanks a lot for your hard work! I just accepted you PR. However, I only realized later that you made the PR against my "master" branch. I suppose it should be against my "connect_astro" branch instead?

Yes I will discuss with Jessica for a illustrative figure. Regarding the subdirectory issue, I think I can avoid the subdirectories, while still showing 2 plots together to the user with plt.show(). I will do this later if there is no other suggestions.

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+
+    Available tripartite rules::
+
+     - 'tripartite_bernoulli_with_pool'
```
```suggestion
- ``tripartite_bernoulli_with_pool``
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    **Synapse specifications (syn_specs)**
+
+    Synapse specifications for tripartite connections are given as a dictionary with specifications
+    for each of the three connections to be created
```
```suggestion
    for each of the three connections to be created::
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
````diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    Synapse specifications for tripartite connections are given as a dictionary with specifications
+    for each of the three connections to be created
+
+    ```
````
```suggestion
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
````diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    {"primary": <syn_spec>,
+     "third_in": <syn_spec>,
+     "third_out": <syn_spec>}
+    ```
````
```suggestion
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
````diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+     "third_out": <syn_spec>}
+    ```
+
+    Here, `"primary"` marks the synapse specification for the connections between `pre` and `post` neurons,
````
```suggestion
    Here, ``"primary"`` marks the synapse specification for the connections between ``pre`` and ``post`` neurons,
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    `third` and `post` neurons.
+
+    Each `<syn_spec>` entry can be any entry that would be possible as synapse specification
+    in a normal `Connect()` call. Any missing entries default to `static_synapse`. If no `syn_specs` argument
```
```suggestion
    in a normal ``Connect()`` call. Any missing entries default to ``static_synapse``. If no ``<syn_spec>`` argument
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
````diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    ```
+
+    Here, `"primary"` marks the synapse specification for the connections between `pre` and `post` neurons,
+    `"third_in"` for connections between `pre` and `third` neurons and `"third_out"` for connections between
````
```suggestion
    ``"third_in"`` for connections between ``pre`` and ``third`` neurons and ``"third_out"`` for connections between
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+
+     - 'tripartite_bernoulli_with_pool'
+
+    See :ref:`tripartite_connectivity` for more details and :doc:`astrocyte_small_network`
```
```suggestion
    See :ref:`tripartite_connectivity` for more details and :doc:`/auto_examples/astrocyte_small_network`
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    -----
+    **Connectivity specification (conn_spec)**
+
+    Available tripartite rules::
```
```suggestion
    Available tripartite rules:
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+    `"third_in"` for connections between `pre` and `third` neurons and `"third_out"` for connections between
+    `third` and `post` neurons.
+
+    Each `<syn_spec>` entry can be any entry that would be possible as synapse specification
```
```suggestion
    Each ``<syn_spec>`` entry can be any entry that would be possible as synapse specification
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+
+    Each `<syn_spec>` entry can be any entry that would be possible as synapse specification
+    in a normal `Connect()` call. Any missing entries default to `static_synapse`. If no `syn_specs` argument
+    is given at all, all three entries default to `static_synapse`.
```
```suggestion
    is given at all, all three entries default to ``static_synapse``.
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+     - 'tripartite_bernoulli_with_pool'
+
+    See :ref:`tripartite_connectivity` for more details and :doc:`astrocyte_small_network`
+    and :doc:`astrocyte_brunel` for examples.
```
```suggestion
    and :doc:`/auto_examples/astrocyte_brunel` for examples.
```

jessica-mitchell commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -293,6 +294,120 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
+
+    Here, `"primary"` marks the synapse specification for the connections between `pre` and `post` neurons,
+    `"third_in"` for connections between `pre` and `third` neurons and `"third_out"` for connections between
+    `third` and `post` neurons.
```
```suggestion
    ``third`` and ``post`` neurons.
```

jessica-mitchell commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,86 @@ must be ``True``.
+
+.. code-block:: python
+
+    A = nest.Create('aeif_cond_alpha_astro', 10)
```
I think all example code blocks should use the double quotes for strings, since this is what our Python formatter expects (Black) in all our Python code. It's not a huge deal, but it would be nice to be consistent.

jessica-mitchell commented:
> @jessica-mitchell Could you review the documentation and examples? One point I am unsure about is that the [astrocyte_small_network.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_small_network.py) and the [astrocyte_brunel.py](https://github.com/nest/nest-simulator/blob/dd15cef8689e5d1f225edf1707d96c1a3fc2e9be/pynest/examples/astrocyte_brunel.py) write figures to files in subdirectories they create. This does not seem to be in line with normal practice for examples. Should they display figures with `plt.show()`? And will all work if a script produces more than one figure?

There are a couple of examples I think that create files rather than use plt.show(). I think so long as it is clear that an image was generated that should be fine (so when running the script, there should be a message about the images being created). 

On another note, I think the astrocyte examples should all be contained within their own subdirectory `pynest/examples/astrocytes/`. Each subfolder under `examples` requires a README.rst for Sphinx-gallery to find and generate the examples, and the paths would need to be updated as well. 

> @HanjiaJiang and @jessica-mitchell I added an explanation of the tripartite rule to `connection_management.rst`, but I think the discussion of the pools would be much clearer with a figure. Could you to collaborate on getting an illustrative figure in?

We have meeting planned to discuss this.

HanjiaJiang commented:
Hi @heplesser @jessica-mitchell I have pushed all new contents that I consider necessary for now. Please have a look, thanks!

Specifically, I added a figure to illustrate the ``pool_type``, as HEP suggested. I also changed or expanded some text in the description created by HEP (e.g. Sacha and I prefer to use ``C`` for the third population, instead of ``T``, to avoid confusion with "target"). But please let me know if they look good to you, or anything still to be improved. Thanks a lot!

heplesser reviewed (changes requested):
@HanjiaJiang This looks very good. I just have two suggestions for the documentation, especially the figure.

HanjiaJiang commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,105 @@ must be ``True``.
+    :align: center
+    :alt: Astrocyte pool type.
+
+    This figure illustrates possible outcomes of connectivity with the two
```
> @HanjiaJiang Below a small suggestion. The line breaks should be re-done once it is intergrated.
> 
> Could you also add the third case for the pool to the illustation, namely a block pool with pool_size > 1, where we have n neurons from B per neuron in C?

Thanks for this good suggestion. I just added a third case for a block pool with ``pool_size`` = 2 and ``n(C)/n(B)`` = 2. I hope it looks good.

I noticed that the other figures in the same page are set with `image::`. It looks a bit weird now if I use `figure::`. So, I changed it back to `image::` and removed the `:alt:` line for now. I hope this is fine.

heplesser replied:
@jessica-mitchell Do you have any comments on the `image::` vs `figure::` question?

heplesser reviewed (approved):
@HanjiaJiang Thanks a lot, approved!

heplesser commented:
@jhnnsnk Ping :).

jessica-mitchell commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+created with probability ``p_primary``, and these connections are
+called "primary" connections. For each primary connection, a
+third-party connection pair involving a node from ``C`` (a third
+``NodeCollections``) is created with the conditional probability
```
```suggestion
``NodeCollection``) is created with the conditional probability
```

jessica-mitchell commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+third-party connection pair involving a node from ``C`` (a third
+``NodeCollections``) is created with the conditional probability
+``p_third_if_primary``. This connection pair includes a connection
+from the node from ``A`` (i.e. the source) to the node from ``C`` in
```
```suggestion
from the node from ``A`` (i.e., the source) to the node from ``C`` in
```

jessica-mitchell commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+``p_third_if_primary``. This connection pair includes a connection
+from the node from ``A`` (i.e. the source) to the node from ``C`` in
+question, and a connection from this node from ``C`` to the node from
+``B`` (i.e. the target). The node from ``C`` to connect to is chosen
```
```suggestion
``B`` (i.e., the target). The node from ``C`` to connect to is chosen
```

jessica-mitchell reviewed (approved):
@HanjiaJiang Looks good, I found a couple of very minor typos, but I approve. Thanks for all your work. As for the image I think it's fine using the image directive. The figure directive is useful for referencing, but I don't think it's necessary.

HanjiaJiang commented:
Thanks a lot. I have merged Jessica's corrections. I will wait and see if Johanna has any suggestions/corrections. :)

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+where the individual ``syn_spec`` elements follow the same rules as
+for the :py:func:`.Connect` function. Any of the three elements can be
+left out, in which case the synapse specification defaults to
+``"static_synapse"``. The ``"primary"`` synapse specification applies
```
Here and below: Use single quotation marks for consistency, e.g., 'primary'.

HanjiaJiang replied:
This is fixed.

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+		      'p_third_if_primary': 0.5,
+                      'pool_type': 'random',
+		      'pool_size': 3}
+    syn_specs_dict = {'third_out': 'sic_connection'}
```
```suggestion
    syn_specs = {'third_out': 'sic_connection'}
```

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+                      'pool_type': 'random',
+		      'pool_size': 3}
+    syn_specs_dict = {'third_out': 'sic_connection'}
+    nest.TripartiteConnect(A, B, C, conn_spec_dict, syn_specs_dict)
```
```suggestion
    nest.TripartiteConnect(A, B, C, conn_spec, syn_specs)
```

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+    A = nest.Create('aeif_cond_alpha_astro', 10)
+    B = nest.Create('aeif_cond_alpha_astro', 20)
+    C = nest.Create('astrocyte_lr_1994', 5)
+    conn_spec_dict = {'rule': 'tripartite_bernoulli_with_pool',
```
```suggestion
    conn_spec = {'rule': 'tripartite_bernoulli_with_pool',
```

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+``B``, ``pool_size`` nodes from ``C`` are chosen randomly without
+replacement.
+
+For block pools, two variants exist. Let ``n(B)`` and ``n(C)`` be the number of
```
I suggest to use for population sizes rather ``N_B`` and ``N_C`` because these can also be used as variables.

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+
+.. code-block:: python
+
+    A = nest.Create('aeif_cond_alpha_astro', 10)
```
Change this example so that it corresponds to one of the figure panels, at best the left one.

jhnnsnk commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+    nest.TripartiteConnect(A, B, C, conn_spec_dict, syn_specs_dict)
+
+
+.. image:: ../static/img/tripartite_pool_type.svg
```
I think that it would be easier to understand the figure if it had panel labels.
Also, I suggest to align it a bit more to the graphical notation in the connectivity concepts study. In #2992 the network sketches from this figure will be included in the documentation: https://doi.org/10.1371/journal.pcbi.1010086.g007
See also my other comment above on naming pre=S, post=T, third=A.

HanjiaJiang replied:
@heplesser @jhnnsnk I have made changes according to Johanna's suggestions here. Please have a look, thanks!

I changed the figure to align with the connectivity concepts, and rearranged some text accordingly. I used a new color to distinguish the third population though. I also expanded the example code to three parts to relate to the figure. Let me know if this looks fine, or I can still improve on figure design, text arrangement, etc. Thanks.

jhnnsnk reviewed (commented):
Thanks a lot for all this great work!
I left suggestions and change requests in the code.
In particular I tried to align the notation to the connectivity concepts study and #2992 .

heplesser commented on nestkernel/conn_builder.h:
```diff
@@ -447,6 +454,82 @@ class BernoulliBuilder : public ConnBuilder
+  void connect_() override;
+
+private:
+  //! Provide index of first third-party node to be assigned to pool for given target node
```
`//!` introduces a doxygen comment for the next line, `//!<` a doxygen comment for what is to  the left of the comment. If the comment is short enough to fit on the same line as the commented variable, I think it is more readable to use the `//!<` approach, but if there is not enough space, then the comment should go on the line above with `//!`.

jhnnsnk replied:
Thanks for explaining!

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_brunel.py:
```diff
@@ -0,0 +1,324 @@
+# are used to create the connectivity of the network.
+
+
+def create_astro_network(scale=1.0):
```
Docstring added.

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_brunel.py:
```diff
@@ -0,0 +1,324 @@
+    return nodes_ex, nodes_in, nodes_astro, nodes_noise
+
+
+def connect_astro_network(nodes_ex, nodes_in, nodes_astro, nodes_noise, scale=1.0):
```
Added.

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_brunel.py:
```diff
@@ -0,0 +1,324 @@
+# by lines and shaded areas, respectively.
+
+
+def plot_dynamics(astro_data, neuron_data, start):
```
Added.

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_brunel.py:
```diff
@@ -0,0 +1,324 @@
+    conn_params_e = {
+        "rule": "tripartite_bernoulli_with_pool",
+        "p_primary": network_params["p_primary"] / scale,
+        "p_third_if_primary": network_params["p_third_if_primary"],
```
Actually the creation of ``third_in`` and ``third_out`` connections with ``p_third_if_primary`` depends on the number of ``primary`` connections. So they are scaled already, when ``primary`` is scaled.

jhnnsnk replied:
Right, maybe it would be useful to add a comment here so that others don't wonder.

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_tripartite.py:
I renamed it to "astrocyte_interaction.py". I am not sure about the second suggestion, because the main purpose of this example is merely to show the interaction between two neurons and an astrocyte, with their dynamics. This file was merged in the previous astrocyte PR, and I forgot why it is involved here, maybe there are some changes... But yes it is a very good suggestion to change the name to avoid confusion, thanks.

HanjiaJiang commented on doc/htmldoc/synapses/connection_management.rst:
```diff
@@ -256,6 +256,107 @@ must be ``True``.
+the connection rule used.
+
+``conn_spec`` must be provided and must specify a tripartite
+connection rule.
```
I made a change accordingly. Please see if it looks fine, thanks!

HanjiaJiang commented on pynest/examples/astrocytes/astrocyte_brunel.py:
```diff
@@ -0,0 +1,324 @@
+    mm_neuron = nest.Create("multimeter", params={"record_from": ["I_SIC"]})
+    mm_astro = nest.Create("multimeter", params={"record_from": ["IP3", "Ca"]})
+
+    # Run pre-simulation
```
Yes this is done, thanks for the suggestion!

HanjiaJiang commented on testsuite/pytests/test_connect_tripartite_bernoulli.py:
```diff
@@ -0,0 +1,313 @@
+
+
+# adapted from connect_test_base.py
+def setup_network(
```
Fixed.

HanjiaJiang commented:
Thank you guys! Finally concluded this PR :).
Many thanks to Jessica for her review as well.
